File: spring-security-in-action/ssia-ch11-ex1-s2/src/main/java/org/example/ssiach11ex1s2/config/SecurityConfig.java
package org.example.ssiach11ex1s2.config;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.example.ssiach11ex1s2.security.filter.InitialAuthenticationFilter;
import org.example.ssiach11ex1s2.security.filter.JwtAuthenticationFilter;
import org.example.ssiach11ex1s2.security.provider.OtpAuthenticationProvider;
import org.example.ssiach11ex1s2.security.provider.UsernamePasswordAuthenticationProvider;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.authentication.AuthenticationManager;
import org.springframework.security.authentication.ProviderManager;
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter;

@Configuration
@RequiredArgsConstructor
public class SecurityConfig {
    private final JwtAuthenticationFilter jwtAuthenticationFilter;
    private final OtpAuthenticationProvider otpAuthenticationProvider;
    private final UsernamePasswordAuthenticationProvider usernamePasswordAuthenticationProvider;


    @Bean
    protected InitialAuthenticationFilter initialAuthenticationFilter() {
        return new InitialAuthenticationFilter(authenticationManager());
    }

    @Bean
    protected SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
        http.authenticationProvider(otpAuthenticationProvider)
                .authenticationProvider(usernamePasswordAuthenticationProvider)
                .csrf(csrf -> csrf.disable())
                .addFilterAt(initialAuthenticationFilter(), BasicAuthenticationFilter.class)
                .addFilterAfter(jwtAuthenticationFilter, BasicAuthenticationFilter.class)
                .authorizeHttpRequests(auth -> auth.anyRequest().authenticated());

        return http.build();
    }

    @Bean
    public AuthenticationManager authenticationManager() {
        return new ProviderManager(List.of(usernamePasswordAuthenticationProvider, otpAuthenticationProvider));
    }
}

File: spring-security-in-action/ssia-ch09-ex1/src/main/kotlin/org/example/ssiach09ex1/config/SecurityConfig.kt
package org.example.ssiach09ex1.config

import org.example.ssiach09ex1.filter.AuthenticationLoggingFilter
import org.example.ssiach09ex1.filter.RequestValidationFilter
import org.example.ssiach09ex1.filter.StaticKeyAuthenticationFilter
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.www.BasicAuthenticationFilter

@Configuration
class SecurityConfig {

    @Autowired
    private lateinit var staticKeyAuthenticationFilter: StaticKeyAuthenticationFilter

    @Bean
    fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http.httpBasic()

        http
            .addFilterAt(staticKeyAuthenticationFilter, BasicAuthenticationFilter::class.java)
//            .addFilterBefore(RequestValidationFilter(), BasicAuthenticationFilter::class.java)
//            .addFilterAfter(AuthenticationLoggingFilter(), BasicAuthenticationFilter::class.java)
            .authorizeHttpRequests { auth -> auth.anyRequest().permitAll() }

        return http.build()
    }
}